### PR TITLE
use restmapper to check ocp and do not delete version claim

### PR DESCRIPTION
**Pull request by zhiweiyin318**
issue: https://github.com/stolostron/backlog/issues/25185
Signed-off-by: Zhiwei Yin <zyin@redhat.com>

### CODE REVIEW

File: pkg/klusterlet/clusterinfo/clusterinfo_controller.go [hunk 1]
```diff
@@ -329,18 +331,17 @@ func (r *ClusterInfoReconciler) getKubeVendor(product string) (kubeVendor cluste
 	return
 }
 
-func (r *ClusterInfoReconciler) isOpenshift() bool {
-	serverGroups, err := r.KubeClient.Discovery().ServerGroups()
+func (r *ClusterInfoReconciler) isOpenshift() (bool, error) {
```
Comment (zhujian7) [inline pkg/klusterlet/clusterinfo/clusterinfo_controller.go:334]: Is it better to move this func to a common package, so it can be used in common?

Comment (zhiweiyin318) [reply pkg/klusterlet/clusterinfo/clusterinfo_controller.go]: changed to check if OCP via product claim.

File: pkg/klusterlet/clusterinfo/clusterinfo_controller.go [hunk 2]
```diff
@@ -121,6 +115,7 @@ func (r *ClusterInfoReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 			newStatus.ClusterID = value
 		case clusterclaim.ClaimOCMProduct:
 			newStatus.KubeVendor = r.getKubeVendor(value)
+			product = value
```
Comment (zhujian7) [inline pkg/klusterlet/clusterinfo/clusterinfo_controller.go:118]: I am not quite sure using this claim is the best approach, I mean if there is a new supported product(ex: microshift maybe), we should add the new product type in https://github.com/stolostron/multicloud-operators-foundation/pull/526/files#diff-ac6674a4f427b422665e2fbea90ad4e776f2946ae22cc9974e4eb9045d5a4312R452 right? and I think it's a little bit error-prone(easy to forget to change).

Is there always a claim to indicate it is OCP no matter what the OCP product is? `id.openshift.io` or `version.openshift.io`?

File: pkg/klusterlet/clusterinfo/clusterinfo_controller.go [hunk 3]
```diff
@@ -130,6 +125,13 @@ func (r *ClusterInfoReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		}
 	}
 
+	// Get distribution info
+	newStatus.DistributionInfo, err = r.getDistributionInfo(ctx, product)
```
Comment (zhujian7) [inline pkg/klusterlet/clusterinfo/clusterinfo_controller.go:129]: Line 123 will be dead code. https://github.com/stolostron/multicloud-operators-foundation/pull/526/files#diff-ac6674a4f427b422665e2fbea90ad4e776f2946ae22cc9974e4eb9045d5a4312R123

**Review by zhujian7 — CHANGES_REQUESTED** [review]

File: pkg/klusterlet/clusterinfo/clusterinfo_controller.go [hunk 4]
```diff
@@ -124,12 +119,17 @@ func (r *ClusterInfoReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
 		case clusterclaim.ClaimOpenshiftVersion:
-			if newStatus.DistributionInfo.Type == clusterv1beta1.DistributionTypeOCP {
-				newStatus.DistributionInfo.OCP.Version = value
-			}
+			ocpVersion = value
 		}
 	}
 
+	// Get distribution info
+	newStatus.DistributionInfo, err = r.getDistributionInfo(ctx, product, ocpVersion)
```
Comment (zhujian7) [inline pkg/klusterlet/clusterinfo/clusterinfo_controller.go:127]: the product will always be empty?

**Review by zhujian7 — CHANGES_REQUESTED** [review]

**Pull request merged**